modules/wallet/src/main/scala/org/tessellation/wallet/Main.scala

```diff
@@ -134,4 +148,37 @@ object Main
         cfg.storepass.coerce.value.toCharArray,
         cfg.keypass.coerce.value.toCharArray
       )
+
+  private def writeJson[F[_]: Async: Console, A: Encoder](outputPath: Option[Path])(a: A): F[Unit] =
+    outputPath.fold(Console[F].println(a.asJson.noSpaces))(writeToJsonFile(_)(a))
+
+  private def mergeMessages[F[_]: Async: Hasher: SecurityProvider](files: NonEmptyList[Path]): F[Signed[CurrencyMessage]] = {
+    type SignedMessages = NonEmptyList[Signed[CurrencyMessage]]
+    type SignedMessagesErrorOr[A] = ValidatedNec[String, A]
+
+    val signedValidator = SignedValidator.make[F]
+
+    def validateMessageValues(inputs: SignedMessages): SignedMessagesErrorOr[SignedMessages] =
+      Validated.condNec(
+        inputs.toIterable.map(_.value).toSet.size == 1,
+        inputs,
+        "Messages are not identical"
+      )
+
+    def validateProofs(inputs: SignedMessages): F[SignedMessagesErrorOr[SignedMessages]] =
+      inputs
+        .traverse(input => signedValidator.validateSignatures(input))
+        .map(_.reduceLeft(_ *> _))
+        .map(_.leftMap(_.map(_.show)).as(inputs))
+
+    files.traverse { p =>
+      OptionT(readFromJsonFile[F, Signed[CurrencyMessage]](p))
+        .getOrRaise(new Exception(s"Unable to load JSON from $p"))
+    }
+      .flatMap(inputs => validateProofs(inputs).map(_ *> validateMessageValues(inputs)))
+      .flatMap {
+        case Valid(inputs)   => Signed(inputs.head.value, inputs.map(_.proofs).reduceLeft(_ ++ _)).pure[F]
```

Comment:
I would just think whether sorting signatures would make sense so that by executing in different order of files or on different machines we do get the same output with the same inputs.

Reply:
Great, I saw an `NonEmptyList` somewhere and thought it was for the sigs as well.